crates/rollup-boost/src/flashblocks/service.rs

```diff
@@ -0,0 +1,318 @@
+    current_payload_id: Arc<RwLock<PayloadId>>,
+
+    // flashblocks payload being constructed
+    best_payload: Arc<RwLock<FlashblockBuilder>>,
+
+    // websocket publisher for sending valid preconfirmations to clients
+    ws_pub: Arc<WebSocketPublisher>,
+}
+
+impl FlashblocksService {
+    pub fn new(client: RpcClient, outbound_addr: String) -> eyre::Result<Self> {
+        let ws_pub = WebSocketPublisher::new(outbound_addr.parse().unwrap())?.into();
+
+        Ok(Self {
+            client,
+            current_payload_id: Arc::new(RwLock::new(PayloadId::default())),
+            best_payload: Arc::new(RwLock::new(FlashblockBuilder::new())),
+            ws_pub,
+        })
+    }
+
+    pub async fn get_best_payload(
+        &self,
+        version: PayloadVersion,
+    ) -> Result<Option<OpExecutionPayloadEnvelope>, FlashblocksError> {
+        // consume the best payload and reset the builder
+        let payload = {
+            let mut builder = self.best_payload.write().await;
+            std::mem::take(&mut *builder).into_envelope(version)?
+        };
+        *self.best_payload.write().await = FlashblockBuilder::new();
+
+        Ok(Some(payload))
+    }
+
+    pub async fn set_current_payload_id(&self, payload_id: PayloadId) {
+        tracing::debug!(message = "Setting current payload ID", payload_id = %payload_id);
+        *self.current_payload_id.write().await = payload_id;
+    }
+
+    async fn on_event(&mut self, event: FlashblocksEngineMessage) {
+        match event {
+            FlashblocksEngineMessage::FlashblocksPayloadV1(payload) => {
+                tracing::debug!(
+                    message = "Received flashblock payload",
+                    payload_id = %payload.payload_id,
+                    index = payload.index
+                );
+
+                // make sure the payload id matches the current payload id
+                if *self.current_payload_id.read().await != payload.payload_id {
+                    error!(message = "Payload ID mismatch",);
+                    return;
+                }
+
+                if let Err(e) = self.best_payload.write().await.extend(payload.clone()) {
+                    error!(message = "Failed to extend payload", error = %e);
+                } else {
+                    // Broadcast the valid message
+                    if let Err(e) = self.ws_pub.publish(&payload) {
+                        error!(message = "Failed to broadcast payload", error = %e);
+                    }
+                }
+            }
+        }
+    }
+
+    pub async fn run(&mut self, mut stream: mpsc::Receiver<FlashblocksPayloadV1>) {
+        loop {
+            let event = stream.recv().await;
+            match event {
+                Some(event) => {
+                    self.on_event(FlashblocksEngineMessage::FlashblocksPayloadV1(event))
+                        .await
+                }
+                None => break,
+            }
+        }
+    }
```

Comment:
Consider simplifying to

```suggestion
 while let Some(event) = stream.recv().await {
        self.on_event(FlashblocksEngineMessage::FlashblocksPayloadV1(event)).await;
    }
```